sonar-slang-plugin/src/main/java/com/sonarsource/slang/kotlin/KotlinTreeVisitor.java

```diff
@@ -114,88 +114,66 @@ public KotlinTreeVisitor(PsiFile psiFile, TreeMetaDataProvider metaDataProvider)
-      return createFunctionDeclarationTree((KtFunction) element, metaData);
+      return createFunctionDeclarationTree(metaData, (KtFunction) element);
     } else if (element instanceof KtIfExpression) {
-      KtIfExpression ifElement = (KtIfExpression) element;
-      Tree condition = createElement(ifElement.getCondition());
-      Tree thenBranch = createElement(ifElement.getThen());
-      Tree elseBranch = createElement(ifElement.getElse());
-      return new IfTreeImpl(metaData, condition, thenBranch, elseBranch);
+      return createIfTree(metaData, (KtIfExpression) element);
     } else if (element instanceof KtWhenExpression) {
-      KtWhenExpression whenElement = (KtWhenExpression) element;
-      // FIXME subjectExpression could be null
-      Tree subjectExpression = createElement(whenElement.getSubjectExpression());
-      List<MatchCaseTree> whenExpressions = list(whenElement.getEntries().stream()).stream()
-        .map(MatchCaseTree.class::cast)
-        .collect(Collectors.toList());
-      return new MatchTreeImpl(metaData, subjectExpression, whenExpressions);
+      return createMatchTree(metaData, (KtWhenExpression) element);
     } else if (element instanceof KtWhenEntry) {
-      KtWhenEntry whenElement = (KtWhenEntry) element;
-      Tree conditions = null;
-      Tree body = createElement(whenElement.getExpression());
-      if (!whenElement.isElse()) {
-        List<Tree> conditionsList = list(Arrays.stream(whenElement.getConditions()));
-        TextPointer startPointer = conditionsList.get(0).metaData().textRange().start();
-        TextPointer endPointer = conditionsList.get(conditionsList.size() - 1).metaData().textRange().end();
-        TextRange textRange = new TextRangeImpl(startPointer, endPointer);
-        TreeMetaData treeMetaData = metaDataProvider.metaData(textRange);
-        conditions = new NativeTreeImpl(treeMetaData, new KotlinNativeKind(KtWhenCondition.class), conditionsList);
-      }
-      return new MatchCaseTreeImpl(getTreeMetaData(whenElement), conditions, body);
-    } else if (element instanceof KtLiteralStringTemplateEntry || element instanceof KtEscapeStringTemplateEntry
-      || (element instanceof KtStringTemplateExpression && !((KtStringTemplateExpression) element).hasInterpolation())) {
+      return createMatchCase(metaData, (KtWhenEntry) element);
+    } else if (isLiteral(element)) {
       return new LiteralTreeImpl(metaData, element.getText());
     } else if (element instanceof KtOperationExpression) {
-      return createOperationExpression((KtOperationExpression) element, metaData);
-    } else if (element instanceof KtParameter) {
-      return new IdentifierTreeImpl(metaData, ((KtParameter) element).getName());
+      return createOperationExpression(metaData, (KtOperationExpression) element);
+    } else if (element instanceof KtParameter && ((KtParameter) element).getName() != null) {
+      return createIdentifierTree(metaData, ((KtParameter) element).getName());
```

Comment:
That seems wrong to me. A parameter is often more than an identifier, it can also have a type. Marcelo wants to add a `ParameterTree` to slang API for the purpose of his rule.

Reply:
Ok. I leave it as a native tree for now.